typescript/cli/src/send/transfer.ts

```diff
@@ -0,0 +1,175 @@
+import assert from 'assert';
+import { ethers } from 'ethers';
+
+import {
+  ERC20__factory,
+  HypERC20App,
+  HypERC20Collateral__factory,
+} from '@hyperlane-xyz/hyperlane-token';
+import {
+  ChainName,
+  HyperlaneContractsMap,
+  HyperlaneCore,
+  MultiProvider,
+} from '@hyperlane-xyz/sdk';
+import { types, utils } from '@hyperlane-xyz/utils';
+
+import { readDeploymentArtifacts } from '../configs.js';
+import { MINIMUM_TEST_SEND_BALANCE } from '../consts.js';
+import { getDeployerContext, getMergedContractAddresses } from '../context.js';
+import { runPreflightChecks } from '../deploy/utils.js';
+import { log, logBlue, logGreen } from '../logger.js';
+import { assertTokenBalance } from '../utils/balances.js';
+
+// TODO improve the UX here by making params optional and
+// prompting for missing values
+export async function sendTestTransfer({
+  key,
+  chainConfigPath,
+  coreArtifactsPath,
+  origin,
+  destination,
+  routerAddress,
+  wei,
+  recipient,
+  timeout,
+}: {
+  key: string;
+  chainConfigPath: string;
+  coreArtifactsPath: string;
+  origin: ChainName;
+  destination: ChainName;
+  routerAddress: types.Address;
+  wei: string;
+  recipient?: string;
+  timeout: number;
+}) {
+  const { signer, multiProvider } = getDeployerContext(key, chainConfigPath);
+  const artifacts = coreArtifactsPath
+    ? readDeploymentArtifacts(coreArtifactsPath)
+    : undefined;
+
+  await assertTokenBalance(
+    multiProvider,
+    signer,
+    origin,
+    routerAddress,
+    wei.toString(),
+  );
+  await runPreflightChecks({
+    local: origin,
+    remotes: [destination],
+    multiProvider,
+    signer,
+    minBalanceWei: MINIMUM_TEST_SEND_BALANCE,
+  });
+
+  await utils.timeout(
+    executeDelivery({
+      origin,
+      destination,
+      routerAddress,
+      wei,
+      recipient,
+      signer,
+      multiProvider,
+      artifacts,
+    }),
+    timeout * 1000,
+    'Timed out waiting for messages to be delivered',
+  );
+}
+
+async function executeDelivery({
+  origin,
+  destination,
+  routerAddress,
+  wei,
+  recipient,
+  multiProvider,
+  signer,
+  artifacts,
+}: {
+  origin: ChainName;
+  destination: ChainName;
+  routerAddress: types.Address;
+  wei: string;
+  recipient?: string;
+  multiProvider: MultiProvider;
+  signer: ethers.Signer;
+  artifacts?: HyperlaneContractsMap<any>;
+}) {
+  const signerAddress = await signer.getAddress();
+  recipient ||= signerAddress;
+
+  const mergedContractAddrs = getMergedContractAddresses(artifacts);
+
+  const core = HyperlaneCore.fromAddressesMap(
+    mergedContractAddrs,
+    multiProvider,
+  );
+
+  const provider = multiProvider.getProvider(origin);
```

Comment:
imo the right place for all this stuff is on the apps

Reply:
I agree in principle but one of my observations from the warp ui dev is that the apps should not handle tx _sending_ only _preparation_. Apps need more fine-grained control over the send flow. So if we want the apps to handle the approve, it needs to be done in an optional way (e.g. separate prepare/send methods)

Reply:
I see
I would have assumed the UI could do everything via the signer on the multiprovider

Reply:
That's a good option for cases where the key is programmatically accessible like here but not so good where txs need to be handed over to wallets (usually via libs in btwn like wagmi) for signing + sending